pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java

```diff
@@ -100,7 +98,15 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
             }
 
             return cb.subscribeAsync();
-        }).collect(Collectors.toList()).stream().map(CompletableFuture::join).collect(Collectors.toList());
+        }).collect(Collectors.toList()).stream().map(consumerFuture -> {
+            try {
+                return consumerFuture.get(consumerOpTimeoutMs, TimeUnit.MILLISECONDS);
```

Comment:
@rdhabalia this sounds like a hack to me. do we know what causes the deadlock here?

Reply:
I don't think it's a hack but hack was made in #3749, where we were interrupting and killing thread without knowing reason. And After checking thread-dump I have found out that function-thread was not completing because it was blocked on this path. 
However, you are correct that we also have to find out root-cause of why consumer-future not was completed and therefore we have to add the timeout. I haven't found anything into the log yet. But I found out that function-worker code-path has many blocking calls and same time it deals with many locks so, it's good practice to avoid forever-blocking call to give a chance to break a deadlock. Therefore, we have added a timeout here.

Reply:
wouldn't be possible to use `CompletableFuture.allOf(listConsumersFuture).get(timeout)` so we could check for `InterruptedException` outside `map()` and throw some pulsar checked exception instead of an unchecked exception ?
I think is better to avoid side effects on lambdas expressions.

Reply:
+1 to @lovelle 's suggestion

Reply:
yes..good point..fixed it.